core/src/main/scala-3/org/http4s/syntax/LiteralsSyntax.scala

```diff
@@ -0,0 +1,80 @@
+ * limitations under the License.
+ */
+
+package org.http4s
+package syntax
+
+import scala.quoted._
+
+trait LiteralsSyntax {
+  extension (inline ctx: StringContext) {
+    inline def uri(args: Any*): Uri = ${LiteralsSyntax.validateUri('{ctx}, '{args})}
+    inline def path(args: Any*): Uri.Path = ${LiteralsSyntax.validatePath('{ctx}, '{args})}
+    inline def scheme(args: Any*): Uri.Scheme = ${LiteralsSyntax.validateUriScheme('{ctx}, '{args})}
+    inline def ipv4(args: Any*): Uri.Ipv4Address = ${LiteralsSyntax.validateIpv4('{ctx}, '{args})}
+    inline def ipv6(args: Any*): Uri.Ipv6Address = ${LiteralsSyntax.validateIpv6('{ctx}, '{args})}
+    inline def mediaType(args: Any*): MediaType = ${LiteralsSyntax.validateMediatype('{ctx}, '{args})}
+    inline def qValue(args: Any*): QValue = ${LiteralsSyntax.validateQvalue('{ctx}, '{args})}
+  }
+}
+
+private[syntax] object LiteralsSyntax {
+
+  trait Validator[A] {
+    def validate(s: String)(using Quotes): ParseResult[Expr[A]]
+  }
+
+  def validateUri(strCtxExpr: Expr[StringContext], argsExpr: Expr[Seq[Any]])(using Quotes) =
+    validate(s => Uri.fromString(s).map(value => Expr(value)), strCtxExpr, argsExpr)
+  def validateUriScheme(strCtxExpr: Expr[StringContext], argsExpr: Expr[Seq[Any]])(using Quotes) =
+    validate(s => Uri.Scheme.fromString(s).map(value => Expr(value)), strCtxExpr, argsExpr)
+  def validatePath(strCtxExpr: Expr[StringContext], argsExpr: Expr[Seq[Any]])(using Quotes) =
+    validate(s => Uri.fromString(s).map(value => Expr(value.path)), strCtxExpr, argsExpr)
+  def validateIpv4(strCtxExpr: Expr[StringContext], argsExpr: Expr[Seq[Any]])(using Quotes) =
+    validate(s => Uri.Ipv4Address.fromString(s).map(value => Expr(value)), strCtxExpr, argsExpr)
+  def validateIpv6(strCtxExpr: Expr[StringContext], argsExpr: Expr[Seq[Any]])(using Quotes) =
+    validate(s => Uri.Ipv6Address.fromString(s).map(value => Expr(value)), strCtxExpr, argsExpr)
+  def validateMediatype(strCtxExpr: Expr[StringContext], argsExpr: Expr[Seq[Any]])(using Quotes) =
+    validate(s => MediaType.parse(s).map(value => Expr(value)), strCtxExpr, argsExpr)
+  def validateQvalue(strCtxExpr: Expr[StringContext], argsExpr: Expr[Seq[Any]])(using Quotes) =
+    validate(qvalue, strCtxExpr, argsExpr)
+
+  def validate[A](validator: Validator[A], strCtxExpr: Expr[StringContext], argsExpr: Expr[Seq[Any]])(using Quotes): Expr[A] = {
+    val sc = strCtxExpr.unliftOrError
+    validate(validator, sc.parts, argsExpr)
+  }
+
+  private def validate[A](validator: Validator[A], parts: Seq[String], argsExpr: Expr[Seq[Any]])(using Quotes): Expr[A] = {
+    if (parts.size == 1) {
+      val literal = parts.head
+      validator.validate(literal) match {
+        case Left(err) =>
+          report.error(err.message)
+          ???
+        case Right(expr) => expr
+      }
+    } else {
+      report.error("interpolation not supported", argsExpr)
+      ???
+    }
+  }
+
+  object qvalue extends Validator[QValue] {
```

Comment:
I cant get this to compile for some reason. Might this have something to do with private constructor and `AnyVal` ?